ITensorGaussianMPS/src/gmps.jl

```diff
@@ -391,50 +812,110 @@ function interleave(xs...)
+  interleaved_inds = interleave(first_half, second_half)
+  ordered_inds = sortperm(interleaved_inds)
+  return M[ordered_inds, ordered_inds]
+end
+
+
 function correlation_matrix_to_mps(
   s::Vector{<:Index},
-  Λ_up::AbstractMatrix,
-  Λ_dn::AbstractMatrix;
+  Λ_up0::AbstractSymmetry,
+  Λ_dn0::AbstractSymmetry;
   eigval_cutoff::Float64=1e-8,
-  maxblocksize::Int=min(size(Λ_up, 1), size(Λ_dn, 1)),
+  maxblocksize::Int=min(size(Λ_up0, 1), size(Λ_dn0, 1)),
   kwargs...,
 )
-  @assert size(Λ_up, 1) == size(Λ_up, 2)
-  @assert size(Λ_dn, 1) == size(Λ_dn, 2)
-  N_up = size(Λ_up, 1)
-  N_dn = size(Λ_dn, 1)
-  @assert N_up == N_dn
-  # Total number of fermion sites
+  @show typeof(Λ_up0)
+  MPS_Elt = promote_type(eltype(Λ_up0.data),eltype(Λ_dn0.data))
+  Λ_up = maybe_drop_pairing_correlations(Λ_up0)
+  Λ_dn = maybe_drop_pairing_correlations(Λ_dn0)
+  @assert size(Λ_up.data, 1) == size(Λ_up.data, 2)
+  @assert size(Λ_dn.data, 1) == size(Λ_dn.data, 2)
+  
+  if ! ((typeof(Λ_up) <: ConservesNfParity && typeof(Λ_dn) <: ConservesNfParity) ||  (typeof(Λ_up) <: ConservesNf && typeof(Λ_dn) <: ConservesNf))
+    error("Λ_up and Λ_dn have incompatible subtypes of AbstractSymmetry")
+  end
+  N_up = div(size(Λ_up.data, 1),site_stride(Λ_up))
+  N_dn = div(size(Λ_dn.data, 1),site_stride(Λ_up))
   N = N_up + N_dn
   ns_up, C_up = correlation_matrix_to_gmps(
     Λ_up; eigval_cutoff=eigval_cutoff, maxblocksize=maxblocksize
   )
   ns_dn, C_dn = correlation_matrix_to_gmps(
     Λ_dn; eigval_cutoff=eigval_cutoff, maxblocksize=maxblocksize
   )
-  # map the up spins to the odd sites and the even spins to the even sites
   C_up = mapindex(n -> 2n - 1, C_up)
   C_dn = mapindex(n -> 2n, C_dn)
-  C = Circuit(interleave(C_up.rotations, C_dn.rotations))
-  ns = interleave(ns_up, ns_dn)
+  C_up_rot=set_data(Λ_up,C_up.rotations)
+  C_dn_rot=set_data(Λ_dn,C_dn.rotations)
+  ns_up=set_data(Λ_up,ns_up)
+  ns_dn=set_data(Λ_dn,ns_dn)
+  C=Circuit(interleave(C_up_rot,C_dn_rot).data)
+  ns=interleave(ns_up,ns_dn).data
   if all(hastags("Fermion"), s)
-    @assert length(s) == N
-    U = [ITensor(s, g) for g in reverse(C.rotations)]
-    ψ = MPS(s, n -> round(Int, ns[n]) + 1, U; kwargs...)
+    U = itensors(s, set_data(Λ_up, C))
+    ψ = MPS(MPS_Elt, s, n -> round(Int, ns[site_stride(Λ_up) * n]) + 1)
+    ψ = apply(U, ψ; kwargs...)
   elseif all(hastags("Electron"), s)
+    ###ToDo: Not sure what to do here yet. Just copied code from old interface below. 
     @assert length(s) == N_up
     @assert length(s) == N_dn
     if isspinful(s)
-      space_up = [QN(("Nf", 0, -1), ("Sz", 0)) => 1, QN(("Nf", 1, -1), ("Sz", 1)) => 1]
-      space_dn = [QN(("Nf", 0, -1), ("Sz", 0)) => 1, QN(("Nf", 1, -1), ("Sz", -1)) => 1]
+      if typeof(Λ_up)<:ConservesNf
+        space_up = [QN(("Nf", 0, -1), ("Sz", 0)) => 1, QN(("Nf", 1, -1), ("Sz", 1)) => 1]
+        space_dn = [QN(("Nf", 0, -1), ("Sz", 0)) => 1, QN(("Nf", 1, -1), ("Sz", -1)) => 1]
+      elseif typeof(Λ_up)<:ConservesNfParity
```

Comment:
Is this correct?

Reply:
That looks ok, though the modules should be `-2` to indicate it is fermionic. You can use the `ITensors.space` functions to check the internal definitions we have:
```julia
julia> space(SiteType("Fermion"))
2

julia> space(SiteType("Fermion"); conserve_qns=true)
2-element Vector{Pair{QN, Int64}}:
 QN("Nf",0,-1) => 1
 QN("Nf",1,-1) => 1

julia> space(SiteType("Fermion"); conserve_nfparity=true)
2-element Vector{Pair{QN, Int64}}:
 QN("NfParity",0,-2) => 1
 QN("NfParity",1,-2) => 1
```